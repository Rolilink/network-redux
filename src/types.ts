export enum NetworkRequestStatusEnum {
  IN_PROGRESS = 'in-progress',
  SUCCESS = 'success',
  ERROR = 'error',
  INVALIDATED = 'invalidated', // an invalidated request no longer have relevance for the current UI
}

export enum NetworkRequestMethodsEnum {
  GET = 'get',
  POST = 'post',
  PUT = 'put',
  DELETE = 'delete',
  PATCH = 'patch',
  OPTIONS = 'options',
  CONNECT = 'connect',
  HEAD = 'head',
  TRACE = 'trace',
}

export interface INetworkRequestError extends Error {
  code?: string,
}

export interface INetworkRequest {
  id: string,
  status: NetworkRequestStatusEnum,
  url: string,
  method: NetworkRequestMethodsEnum,
  query?: any,
  body?: any,
  meta?: any,
  startedAt: Date,
  finishedAt?: Date,
  error?: INetworkRequestError,
}
