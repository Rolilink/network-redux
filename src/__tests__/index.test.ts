import { makeReduxNetwork } from '..';
import { INetworkRequest, NetworkRequestStatusEnum, NetworkRequestMethodsEnum } from '../types';

const requestA: INetworkRequest = {
  id: 'a',
  // eslint-disable-next-line no-undef
  status: NetworkRequestStatusEnum.IN_PROGRESS,
  // eslint-disable-next-line no-undef
  method: NetworkRequestMethodsEnum.GET,
  url: 'localhost:3000/fakeURL/a',
  startedAt: new Date(1594672653959),
};

const requestB: INetworkRequest = {
  id: 'b',
  // eslint-disable-next-line no-undef
  status: NetworkRequestStatusEnum.IN_PROGRESS,
  // eslint-disable-next-line no-undef
  method: NetworkRequestMethodsEnum.GET,
  url: 'localhost:3000/fakeURL/b',
  startedAt: new Date(1594672653959),
};

describe('makeReduxNetwork', () => {
  it('should return a redux duck', () => {
    const duck = makeReduxNetwork('organization');

    expect(duck.name).toBe('organization/network');
    expect(duck.reducer).toBeDefined();
    expect(duck.actions).toBeDefined();
    expect(duck.selectors).toBeDefined();
    expect(duck.caseReducers).toBeDefined();
  });
});

describe(('makeReduxNetwork.reducer'), () => {
  it('should be able to add a network request', () => {
    const { reducer, actions } = makeReduxNetwork('organization');

    const newState = reducer(undefined, actions.addNetworkRequest(requestA));

    expect(newState).toMatchSnapshot();
  });

  it('should be able to remove a network request', () => {
    const oldState = {
      entities: {
        [requestA.id]: requestA,
        [requestB.id]: requestB,
      },
      ids: [
        'a',
        'b',
      ],
    };
    const { reducer, actions } = makeReduxNetwork('organization');

    const newState = reducer(oldState, actions.removeNetworkRequest(requestA.id));

    expect(newState).toMatchSnapshot();
  });

  it('should be able to update a network request', () => {
    const oldState = {
      entities: {
        [requestA.id]: requestA,
      },
      ids: [
        'a',
      ],
    };
    const { reducer, actions } = makeReduxNetwork('organization');

    const newState = reducer(oldState, actions.updateNetworkRequest({
      id: requestA.id,
      changes: {
        status: NetworkRequestStatusEnum.SUCCESS,
      },
    }));

    expect(newState).toMatchSnapshot();
  });

  it('should be able to remove all network requests', () => {
    const oldState = {
      entities: {
        [requestA.id]: requestA,
        [requestB.id]: requestB,
      },
      ids: [
        'a',
        'b',
      ],
    };
    const { reducer, actions } = makeReduxNetwork('organization');

    const newState = reducer(oldState, actions.removeAllNetworkRequests());

    expect(newState).toMatchSnapshot();
  });
});
