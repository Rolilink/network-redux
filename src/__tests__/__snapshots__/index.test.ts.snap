// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeReduxNetwork.reducer should be able to add a network request 1`] = `
Object {
  "entities": Object {
    "a": Object {
      "id": "a",
      "method": "get",
      "startedAt": 2020-07-13T20:37:33.959Z,
      "status": "in-progress",
      "url": "localhost:3000/fakeURL/a",
    },
  },
  "ids": Array [
    "a",
  ],
}
`;

exports[`makeReduxNetwork.reducer should be able to remove a network request 1`] = `
Object {
  "entities": Object {
    "b": Object {
      "id": "b",
      "method": "get",
      "startedAt": 2020-07-13T20:37:33.959Z,
      "status": "in-progress",
      "url": "localhost:3000/fakeURL/b",
    },
  },
  "ids": Array [
    "b",
  ],
}
`;

exports[`makeReduxNetwork.reducer should be able to remove all network requests 1`] = `
Object {
  "entities": Object {},
  "ids": Array [],
}
`;

exports[`makeReduxNetwork.reducer should be able to update a network request 1`] = `
Object {
  "entities": Object {
    "a": Object {
      "id": "a",
      "method": "get",
      "startedAt": 2020-07-13T20:37:33.959Z,
      "status": "success",
      "url": "localhost:3000/fakeURL/a",
    },
  },
  "ids": Array [
    "a",
  ],
}
`;
